replica/table.cc

```diff
@@ -2955,10 +2957,10 @@ table::mutation_query(schema_ptr s,
 
     if (!saved_querier || (!q.are_limits_reached() && !r.is_short_read())) {
         co_await q.close();
-        querier_opt = {};
+        querier_opt.reset();
     }
-    if (saved_querier) {
-        *saved_querier = std::move(querier_opt);
+    if (saved_querier && querier_opt) {
+        saved_querier->emplace(std::move(*querier_opt));
     }
```

Comment:
Alternative approach: make querier a unique_ptr and (and use optimized_optional, or just unique_ptr's optionalness). De-pointerize the contents of queriers.

Reply:
I'm not sure it's that simple.  Maybe with same massage to the multishard querier.
The root problem is the separation of the readers that may be kept as inactive reads by the reader concurrency semaphore, and state they rely on, like the partition range and slice. Those are kept external to the reader using mechanisms like lifecycle policy.  Some readers do keep some of the state (see the evictable_reader_v2) and in the case of the shard_mutation_querier, the range and slice are kept long term in reader_meta::remote_parts so they can be used for recreating the reader in case it was evicted.
Maybe the querier impl can be kept as a whole in `remote_parts`, but that might conflict with the querier cache.
I wish we had one mechanism for saving the query state that would work for all use cases.

Ah, and we also have `query_state` in table::query that keep partition ranges, but it is not kept long term.